benchmarks/src/main/scala/monix/TaskGatherUnordered.scala

```diff
@@ -0,0 +1,186 @@
+ */
+
+package monix
+
+import java.util.concurrent.TimeUnit
+import monix.eval.{Callback, Task}
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import monix.execution.atomic.Atomic
+import monix.execution.cancelables.{CompositeCancelable, StackedCancelable}
+import org.openjdk.jmh.annotations._
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+import scala.collection.generic.CanBuildFrom
+import scala.collection.TraversableOnce
+import scala.annotation.tailrec
+
+/*
+ * Sample run:
+ *
+ *     sbt "benchmarks/jmh:run -i 10 -wi 10 -f 1 -t 1 monix.TaskGatherUnorderedBenchmark"
+ *
+ * Which means "10 iterations" "5 warmup iterations" "1 fork" "1 thread".
+ * Please note that benchmarks should be usually executed at least in
+ * 10 iterations (as a rule of thumb), but more is better.
+ */
+@State(Scope.Thread)
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class TaskGatherUnorderedBenchmark {
+
+  @Param(Array("10", "1000", "100000"))
+  var taskSize: Int = _
+
+  @Param(Array("10", "500", "10000"))
+  var collSize: Int = _
+
+
+  def sum(n: Int, acc: Long = 0): Task[Long] =
+    if (n == 0) Task.evalAlways(acc)
+    else Task.evalAlways(n).flatMap(x => sum(x-1, acc + x))
+
+  def tasks[M[_] <: TraversableOnce[_]](implicit
+    cbf: CanBuildFrom[Nothing, Task[Long], M[Task[Long]]]
+  ): M[Task[Long]] = {
+    val builder = cbf.apply()
+    builder.sizeHint(collSize)
+    for (_ <- 0 until collSize) { builder += sum(taskSize) }
+    builder.result()
+  }
+
+
+  @Benchmark
+  def gather_list(): List[Long] = {
+    val task = Task.gather(tasks[List])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def gather_vector(): Vector[Long] = {
+    val task = Task.gather(tasks[Vector])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_blocking_list(): List[Long] = {
+    val task = gatherUnorderedBlocking(tasks[List])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_blocking_vector(): Vector[Long] = {
+    val task = gatherUnorderedBlocking(tasks[Vector])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_nonblocking_list(): List[Long] = {
+    val task = gatherUnorderedNonBlocking(tasks[List])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+  @Benchmark
+  def unordered_nonblocking_vector(): List[Long] = {
+    val task = gatherUnorderedNonBlocking(tasks[Vector])
+    Await.result(task.runAsync, Duration.Inf)
+  }
+
+
+  def gatherUnorderedBlocking[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])(
+    implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]
+  ): Task[M[A]] = {
+    Task.unsafeCreate { (scheduler, conn, finalCallback) =>
+      val lock = new Object()
+      val builder = cbf(in)
+      var remaining = in.size
+
+      val composite = CompositeCancelable()
+      conn.push(composite)
+
+      for (task <- in; if !composite.isCanceled) {
```

Comment:
Not sure. Maybe the `map` has lazy behavior, probably depending on what you give as input. If it has lazy behavior, then it won't execute, so there will be nothing to call the final callback.

But a `foreach` is better in this case anyway.